[Hong Kai] iP

ong-hongkai opened the pull request:
# Winston
> "Be water, my friend."  -Bruce Lee ([source](https://brucelee.com/podcast-blog/2016/7/20/2-be-water-my-friend))

Winston *frees* your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* Fast to use

All you need to do is,
1. ~~ask for the download link~~
2. download it from [here](https://github.com/Imerbear/ip)
2. double click it
3. add tasks
4. let it manage tasks for you :star_struck:

And it is **FREE!**

Features:
- [x] Manages tasks
- [ ] Manages deadlines (Coming soon)
- [ ] Sends reminders (Coming soon)

As an introduction to Java, here's the way to `print` "Hello World!"
``` java
public class Main {
    public static void main(String[] args) {
        System.out.println("Hello World!")
    }
}
```

Commit:
Add Gradle support

ong-hongkai committed:
Chatbot with level-1 features created

ong-hongkai committed:
Added Task class to store task description

ong-hongkai committed:
Added arraylist to store data for list function

ong-hongkai committed:
Added a boolean field to store completion info
Added documentation

ong-hongkai committed:
Added documentation as well as methods to support command "done"

ong-hongkai committed:
Added DeadLine Task

ong-hongkai committed:
Added Event Task class

ong-hongkai committed:
Tweaked Task to allow for children classes

ong-hongkai committed:
Added ToDoTask task class

ong-hongkai committed:
Changed implementation of chatbot

ong-hongkai committed:
Removed Duke File

ong-hongkai committed:
Adding input and Expected results

ong-hongkai committed:
Tweaked runtest.sh to run on seperate file

ong-hongkai committed:
Tweaked output string for toString methods

ong-hongkai committed:
Added Error handling for done method

ong-hongkai committed:
Added Error Handling for non-command inputs

ong-hongkai committed:
Added delete command and size of list feature

ong-hongkai committed:
Added documentation

ong-hongkai committed:
Changed indentations for switch and case

ong-hongkai committed:
Add support for auto-save feature

ong-hongkai committed:
Add auto-save feature

ong-hongkai committed:
Add DateHandler to manage date transformation

ong-hongkai committed:
Change the way date and time is handled

ong-hongkai committed:
Add support for date inputs

ong-hongkai committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/DeadLine.java
#	src/main/java/Event.java
#	src/main/java/Winston.java

ong-hongkai committed:
Fix bugs of deadline not re-asking for proper date format

ong-hongkai committed:
Add Ui Class

ong-hongkai committed:
Add AddDeadLineCommand class

ong-hongkai committed:
Add AddEventCommand class

ong-hongkai committed:
Add AddToDoCommand class

ong-hongkai committed:
Add Command abstract class

ong-hongkai committed:
Add DeleteCommand class

ong-hongkai committed:
Add DoneCommand class

ong-hongkai committed:
Add InvalidCommand class

ong-hongkai committed:
Add Parser class

ong-hongkai committed:
Add ShowListCommand class

ong-hongkai committed:
Add Storage class

ong-hongkai committed:
Add TaskList class

ong-hongkai committed:
Add TerminateCommand class

ong-hongkai committed:
Add support for more OOP to Winston class

ong-hongkai committed:
Add classes to winston package

ong-hongkai committed:
Change method name to better represent method actions

ong-hongkai committed:
Change Ui class method name to better reflect method action

ong-hongkai committed:
Fix bug to return InvalidCommand if date can't be parsed

ong-hongkai committed:
Add Junit Tests for Parser and TaskList

ong-hongkai committed:
Add .jar file

ong-hongkai committed:
Add JavaDocs

ong-hongkai committed:
Add JavaDocs

ong-hongkai committed:
Change code to follow coding standards

ong-hongkai committed:
Add support for find command

Support was added through:
findCommand class: the class is a child class of the abstract parent class Command
Parser: add another else if statement for "find" command
TaskList: add method find that returns a TaskList Object with the found information
Task: add method to determine if task contains given string

ong-hongkai committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/winston/Task.java

ong-hongkai committed:
Merge commit '7874b7982334cf52093d580383c14820cfb22bc2'

# Conflicts:
#	src/main/java/winston/Task.java

ong-hongkai committed:
Add JavaDocs for FindCommand class

ong-hongkai committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/winston/DeadLine.java
#	src/main/java/winston/Event.java
#	src/main/java/winston/Parser.java

ong-hongkai commented on src/main/java/winston/Storage.java:
```diff
@@ -0,0 +1,100 @@
+package winston;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
```
```suggestion
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
```

test

ong-hongkai committed:
Change ordering of import statement

ong-hongkai committed:
Merge branch 'add-gradle-support' of https://github.com/Imerbear/ip into A-Gradle

ong-hongkai committed:
Add Gradle to support automated testing and checkstyle

ong-hongkai committed:
Add dependencies for javaFX

ryantianj commented on src/main/java/Winston.java:
```diff
@@ -0,0 +1,41 @@
+    
+    public static void main(String[] args) {
+        boolean isExit = false;
+        Winston winston1 = new Winston();
```
Try not to use numbers for variables. Perhaps naming a variable with something more meaningful will make it clearer

ryantianj commented on src/main/java/Winston.java:
```diff
@@ -0,0 +1,41 @@
+        Scanner scan = new Scanner(System.in);
+        Ui.welcomeMessage();
+        while (!isExit) {
+            Command c = winston1.parser.parse(scan.nextLine());
```
Try not to use short-forms for variables, perhaps 'command' instead of c would b better here

ryantianj commented on src/main/java/winston/AddTodoCommand.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    /**
+     * Adds a 'todoTask' to the TaskList and prints the number of uncompleted tasks remaining
```
Adds a toDoTask (camel casing)

ryantianj commented on src/main/java/winston/Parser.java:
```diff
@@ -0,0 +1,90 @@
+            }
+        } else {
+            return new InvalidCommand(taskList);
+        }
```
I noticed that the exceptions being thrown resulted in similar catch blocks (return new InvalidCommand(taskList);) Perhaps abstracting the command parts would make the code cleaner

ryantianj commented on src/main/java/winston/Task.java:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Private constructor for Task
+     *
+     * @param description is the string of the description of the given task
```
missing @param for isCompleted

ryantianj reviewed (commented):
Code format was ok in general :). Do note that all javadoc comments should end with a fullstop, including @param, @return etc...

ong-hongkai committed:
Add codebase before combination of GUI and logic

ong-hongkai committed:
Change controller to allow for MainWindow.fxml to find MainWindow class

ong-hongkai committed:
Add a GUI for Winston

ong-hongkai committed:
Change method names to comply with coding standards

dhshah1 commented on src/main/java/winston/DateHandler.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    /**
+     * A method that will convert a date in strong format to LocalDate format
+     * @param str String of the date
```
Maybe an empty line should be added between the description and param section for java docs. There are a few other files where this empty line should be added.

```suggestion
     * A method that will convert a date in strong format to LocalDate format.
     *
     * @param str String of the date.
```

dhshah1 commented on src/main/java/winston/Ui.java:
```diff
@@ -0,0 +1,43 @@
+        return "Invalid date format. Please give a valid date format. E.g 2021-12-12";
+    }
+    
+    public static String welcomeMessage() {
+        return "Hi there! Winston reporting.\nWhat can I do for you?\n" +
+                "Available Commands: done, list, todo, deadline, event, bye, delete, findString";
```
The coding standards recommend splitting into a new line before the operator.

```suggestion
    public static String welcomeMessage() {
        return "Hi there! Winston reporting.\nWhat can I do for you?\n" 
                + "Available Commands: done, list, todo, deadline, event, bye, delete, findString";
```

dhshah1 commented on src/main/java/winston/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,24 @@
+package winston;
+
+/**
+ *  Represents the deadline command from the parent abstract class Command.
```
This is a really minor suggestion: maybe the extra space at the start of the description block can be removed for consistency. This is present for a few of the command classes.
```suggestion
 * Represents the deadline command from the parent abstract class Command.
```

dhshah1 commented on src/main/java/winston/Event.java:
```diff
@@ -0,0 +1,49 @@
+     * @return a string with the type, completion status, task description and due date
+     */
+    @Override
+    public String saveFormat() {
+        return this.type + "," + super.saveFormat() + "," + this.dateAndTime;
+
+    }
```
```suggestion
    public String saveFormat() {
        return this.type + "," + super.saveFormat() + "," + this.dateAndTime;
    }
```

dhshah1 commented on src/main/java/winston/DoneCommand.java:
```diff
@@ -0,0 +1,24 @@
+public class DoneCommand extends Command {
+    private final int index;
+    
+    public DoneCommand(TaskList tasklist, int index) {         
```
Perhaps a javadoc could be added here for consistency. Currently some classes have javadocs for their constructors and some don't.

dhshah1 commented on src/main/java/winston/ShowListCommand.java:
```diff
@@ -0,0 +1,20 @@
+     * Prints the all tasks on TaskList as well as the number of uncompleted tasks remaining
+     */
+    @Override
+    public String run() {
+        return Ui.printList(taskList) + Ui.printTasksLeft(taskList.numberOfIncompleteTasks());
+
+                
+    }
```
```suggestion
    public String run() {
        return Ui.printList(taskList) + Ui.printTasksLeft(taskList.numberOfIncompleteTasks());
    }
```

dhshah1 commented on src/main/java/winston/TerminateCommand.java:
```diff
@@ -0,0 +1,28 @@
+        isTerminated = true;
+    }
+
+    /**
+     * A method that prints the terminating message found in class Ui
+     */
```
Perhaps a @return should be added in the javadoc

dhshah1 commented on src/main/java/winston/ToDoTask.java:
```diff
@@ -0,0 +1,42 @@
+ * A class extension of Task representing a ToDoTask
+ */
+public class ToDoTask extends winston.Task {
+    private final String type;
+    /**
```
```suggestion
    private final String type;
    
    /**
```

dhshah1 commented on src/main/java/winston/Ui.java:
```diff
@@ -0,0 +1,43 @@
+
+/**
+ * A class that manages all lines that handles user experience and lines that are commonly used.
+ */
+public class Ui {
```
Perhaps can consider adding javadocs for the public methods of the Ui class

dhshah1 reviewed (commented):
Overall, I felt that the code was nicely formatted and adhered to most of the coding standards. I left just a few minor suggestions for your consideration relating to spacing, indentation and javadoc formatting.

ong-hongkai commented:
@dhshah1 Thanks for taking the time to review and give suggestions on the way I can improve my code! I really do appreciate it 😄. I shall fix these errors that I have made for the next IP submission.

ntwbruce commented on src/main/java/winston/Command.java:
```diff
@@ -0,0 +1,14 @@
+package winston;
+
+/**
+ *  Represents the commands that will be used during the usage of winston
```
Very small issue: full stop at the end of description.

ntwbruce commented on src/main/java/winston/DeadLine.java:
```diff
@@ -0,0 +1,50 @@
+    public DeadLine(String description, String dueDate, boolean isCompleted) {
+        super(description, isCompleted);
+        this.dueDate = DateHandler.readDate(dueDate);
+        this.type = "D";
```
Minor suggestion: I think you don't need a 'this' keyword if the 'type' field isn't shadowed by a constructor parameter.

ntwbruce commented on src/main/java/winston/FindCommand.java:
```diff
@@ -0,0 +1,23 @@
+package winston;
+
+public class FindCommand extends Command {
+    private String str;
```
Might not be necessary since it's the only String field in this class but could give it a more descriptive name

ntwbruce commented on src/main/java/winston/Task.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class Task {
+    private final String description;
+    private boolean done;
```
Should follow boolean naming convention
```suggestion
    private boolean isDone;
```

ntwbruce reviewed (commented):
Looks good! Just a few minor suggestions.

ong-hongkai committed:
Build .jar file using gradle

ong-hongkai commented:
@ntwbruce Thanks for the suggestions! Have made the necessary changes based on them

ong-hongkai committed:
Fix the code to comply with coding standards

ong-hongkai committed:
Remove "this" keyword when it isn't necessary

ong-hongkai committed:
Improve code quality

Changed method names in TaskList and Storage classes to make method names more intuitive and easier to understand

ong-hongkai committed:
Improve code quality

Overloaded InvalidCommand constructor to allow for error message to be passed, which allows for more specific error message
Cleaned the implementation of parse method in Parser to make it easier to read

ong-hongkai committed:
Comply with java coding standards

ong-hongkai committed:
Simplify expression to improve readability

ong-hongkai committed:
Enable assertion

ong-hongkai committed:
Add Assertions

Assertions are added in DeadlineCommand, AddEventCommand, TodoCommand, DeadLine, DeleteCommand, DoneCommand, Event, FindCommand,
	InvalidCommand, ListCommand, TerminateCommand, ToDoTask, Storage and TaskList
Assertions are used to verify the code to ensure that some of the assumptions are true throughout the runtime of the program

ong-hongkai committed:
Change implementation of findString

Used streams instead of a regular for loop

ong-hongkai committed:
Merge pull request #1 from Imerbear/branch-A-CodeQuality

Merge CodeQuality with master

ong-hongkai committed:
Merge branch 'master' of https://github.com/Imerbear/ip into branch-A-Assertions

# Conflicts:
#	src/main/java/winston/DeadLine.java
#	src/main/java/winston/InvalidCommand.java
#	src/main/java/winston/Storage.java
#	src/main/java/winston/TaskList.java

ong-hongkai committed:
Merge pull request #3 from Imerbear/branch-A-Assertions

Branch a assertions

ong-hongkai committed:
Merge pull request #2 from Imerbear/branch-A-Streams

Change implementation of findString

ong-hongkai committed:
Add gradle.yml

ong-hongkai committed:
Add support for new command Update

ong-hongkai committed:
Add update as part of available commands in welcome message

ong-hongkai committed:
Update CI

ong-hongkai committed:
Remove unnecessary line of code

ong-hongkai committed:
Change GUI

Replaced both profile images as well as improve the look of the application

ong-hongkai committed:
Change chatbot personality

ong-hongkai committed:
Tweak UX

ong-hongkai committed:
Add UserGuide and Ui.png

ong-hongkai committed:
Fix bugs with findstring command

ong-hongkai committed:
Tweak UG

ong-hongkai committed:
Update UG

ong-hongkai committed:
Tweak UG

ong-hongkai committed:
Tweak UG

ong-hongkai committed:
Update UG

ong-hongkai committed:
Tweak UG

ong-hongkai committed:
Fix bug for done and delete commands for not accounting for indexOutOfBoundErrors
